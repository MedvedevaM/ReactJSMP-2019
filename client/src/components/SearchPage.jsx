import React, { Component } from 'react';
import { FilmsContainer } from './FilmsContainer.jsx';
import { FilmSearch } from './FilmSearch.jsx';

export class SearchPage extends Component {
    state = {
        films: [],
        quantityOfFilms: 0,
        chosenFilm: null,
        searchBy: 'Title',
        sortBy: 'rating'
    }
    
    searchFilms = () => {

    }

    onSearchModeClick = () => {
        this.setState({
            chosenFilm: null
        });
    }

    onSearchParameterButtonClick = (event) => {
        if (event.target.tagName === 'BUTTON' && event.target.id) {
            this.setState({
                searchBy: event.target.id
            });
        }; 
    }

    onSortParameterClick = (event) => {
        if (event.target.tagName === 'LI' && event.target.id) {
            this.setState({
                sortBy: event.target.id
            });
        }; 
    }

    chooseFilm = (event) => {
        if (event.target.id) {
            this.setState({
                chosenFilm: this.state.films.find((film) => film.id == event.target.id)
            });
        }
    }
    componentDidMount() {
        fetch('http://reactjs-cdp.herokuapp.com/movies',
            {
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                method: 'GET'
            })
            .then(response => response.json())
            .then((result) => this.setState({ 
                films: result.data,
                quantityOfFilms: result.data.length
            }));
    }
    render() {
        return (
            <>  
                <FilmSearch {...this.state} 
                    searchFilms={this.searchFilms} 
                    onSearchModeClick={this.onSearchModeClick} 
                    onSearchParameterButtonClick={this.onSearchParameterButtonClick} />
                <FilmsContainer {...this.state}
                    chooseFilm={this.chooseFilm}
                    onSortParameterClick={this.onSortParameterClick} />
                <footer>
                    <div>
                        <p className='logo'>netflixroulette</p>
                    </div>
                </footer>
            </>
        );
    }
};