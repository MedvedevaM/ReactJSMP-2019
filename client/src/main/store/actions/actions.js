export const GET_FILMS = 'GET_FILMS';
export const GET_QUANTITY_OF_FILMS = 'GET_QUANTITY_OF_FILMS';
export const SET_SORT_PARAMETER = 'SET_SORT_PARAMETER';
export const SET_SEARCH_PARAMETER = 'SET_SEARCH_PARAMETER';
export const SET_CHOSEN_FILM = 'SET_CHOSEN_FILM';

export const getFilms = (films) => {
  return {
    type: GET_FILMS,
    films
  }
}

export const getQuantityOfFilms = (quantityOfFilms) => {
  return {
    type: GET_QUANTITY_OF_FILMS,
    quantityOfFilms
  }
}

export const setSortParameter = (sortBy) => {
  return {
    type: SET_SORT_PARAMETER,
    sortBy
  }
}

export const setSearchParameter = (searchBy) => {
  return {
    type: SET_SEARCH_PARAMETER,
    searchBy
  }
}

export const setChosenFilm = (chosenFilm) => {
  return {
    type: SET_CHOSEN_FILM,
    chosenFilm
  }
}

// middlewares

export function fetchFilms(url) {
  return (dispatch) => {
    return fetch(url)
      .then((response) => response.json())
      .then((films) => {
        dispatch(getFilms(films.data));
        dispatch(getQuantityOfFilms(films.data.length));
      });
  };
}

export function getSortedFilms(films, parameter) {
  return (dispatch) => {
    if (parameter === "rating") {
      films = films.sort((a, b) => a.vote_average - b.vote_average);
    }
    if (parameter === "release date") {
      films = films.sort((a, b) => a.release_date.slice(0, 4) - b.release_date.slice(0, 4));
    }
    dispatch(getFilms(films));
  };
}

