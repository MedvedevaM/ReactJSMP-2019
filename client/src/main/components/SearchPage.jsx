import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setSearchParameter, setSortParameter, setChosenFilm, fetchFilms, setSearchValue, searchFilms } from '../store/actions/actions';
import { getFilms, getChosenFilm, getFilmsQuantity, getSearchParameter, getSortParameter, getSearchValue, getFoundFilms } from '../store/reducers/selectors';
import FilmsContainer from './FilmsContainer.jsx';
import FilmSearch from './FilmSearch.jsx';

export class SearchPage extends Component {
  componentDidMount() {
    const { fetchFilms } = this.props;
    fetchFilms('http://reactjs-cdp.herokuapp.com/movies');
  }

  searchFilms = (event) => {
    const { films, setSearchValue, searchBy, searchFilms } = this.props;
    setSearchValue(event.target.value);
    searchFilms(event.target.value, films, searchBy);
  };

  onSearchModeClick = () => {
    const { setChosenFilm } = this.props;
    setChosenFilm(null);
  }

  onSearchParameterClick = (event) => {
    const { setSearchParameter } = this.props;
    const searchParameter = event.target.dataset.parameter;
    if (searchParameter) {
      setSearchParameter(searchParameter);
    }
  }

  onSortParameterClick = (event) => {
    const { setSortParameter, films } = this.props;
    const { sortParameter } = event.target.dataset;
    if (sortParameter) {
      setSortParameter(sortParameter);
      getFilms(films, sortParameter);
    }
  }

  chooseFilm = (event) => {
    const { setChosenFilm, films } = this.props;
    const chosenFilmId = event.target.dataset.filmId;
    if (chosenFilmId) {
      const chosenFilm = films.find(film => film.id === +chosenFilmId);
      setChosenFilm(chosenFilm);
    }
  }

  render() {
    return (
      <>
        <FilmSearch
          {...this.props}
          searchFilms={this.searchFilms}
          onSearchModeClick={this.onSearchModeClick}
          onSearchParameterClick={this.onSearchParameterClick}
        />
        <FilmsContainer
          {...this.props}
          chooseFilm={this.chooseFilm}
          onSortParameterClick={this.onSortParameterClick}
        />
        <footer>
          <div>
            <p className="logo common-color-1">netflixroulette</p>
          </div>
        </footer>
      </>
    );
  }
}

export const mapStateToProps = store => ({ films: getFilms(store),
  quantityOfFilms: getFilmsQuantity(store),
  chosenFilm: getChosenFilm(store),
  searchBy: getSearchParameter(store),
  searchValue: getSearchValue(store),
  sortBy: getSortParameter(store),
  foundFilms: getFoundFilms(store) });

export const mapDispatchToProps = dispatch => ({ setSearchParameter: parameter => dispatch(setSearchParameter(parameter)),
  setSearchValue: value => dispatch(setSearchValue(value)),
  setSortParameter: parameter => dispatch(setSortParameter(parameter)),
  setChosenFilm: film => dispatch(setChosenFilm(film)),
  fetchFilms: url => dispatch(fetchFilms(url)),
  getFilms: films => dispatch(getFilms(films)),
  searchFilms: (value, films, searchParameter) => dispatch(searchFilms(value, films, searchParameter)) });

export default connect(mapStateToProps, mapDispatchToProps)(SearchPage);
