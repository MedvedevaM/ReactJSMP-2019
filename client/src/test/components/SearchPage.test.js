import React from "react";
import { shallow, mount} from "enzyme";
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
// import { createStore, applyMiddleware } from 'redux';
// import thunk from 'redux-thunk';
// import { appReducer } from './../../main/store/reducers/reducers';
import { SearchPage } from "../../main/components/SearchPage.jsx";
import { getFilms } from "../../main/store/actions/actions";

// const store = createStore(
//     appReducer,
//     applyMiddleware(thunk)
// );
const films = [{
    id: 1111,
    release_date: "2020-05-01",
    title: "Guardians of the Galaxy Vol. 3",
    vote_average: 8,
    runtime: 100,
    tagline: "Tag line",
    overview: "The third film based on Marvel's Guardians of the Galaxy.",
    poster_path: "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
    genres: ['Drama', 'Adventure']
}, {
    id: 2222,
    release_date: "2020-05-01",
    title: "Guardians of the Galaxy Vol. 3",
    vote_average: 8,
    runtime: 100,
    tagline: "Tag line",
    overview: "The third film based on Marvel's Guardians of the Galaxy.",
    poster_path: "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
    genres: ['Drama', 'Adventure']
}];

describe("search page rendering and functionality", () => {
    it("check radio buttons rendering", () => {
        const searchPage = shallow(<SearchPage /> );
        expect(toJson(searchPage)).toMatchSnapshot();
    });

    it("choosing film", () => {
        


        // jest.spyOn(window, 'fetch').mockImplementation(() => Promise.resolve({
        //     json: () => Promise.resolve(mockData),
        // }));
        const mockProps = {
            films: films,
            quantityOfFilms: 2,
            chosenFilm: null,
            searchBy: "Title",
            sortBy: "release date",
        }

        const searchPage = mount(<SearchPage props={mockProps} />);

        expect(searchPage.props().films).toEqual(mockData.data);
        searchPage.instance().onSortParameterClick({
            target: {
                dataset: {
                    sortParameter: ""
                }
            }
        });
        expect(searchPage.props().sortBy).toEqual("release date");
        searchPage.instance().onSortParameterClick({
            target: {
                dataset: {
                    sortParameter: "rating"
                }
            }
        });
        expect(searchPage.props().sortBy).toEqual("rating");
        searchPage.instance().onSearchParameterClick({
            target: {
                dataset: {
                    parameter: ""
                }
            }
        });
        expect(searchPage.props().searchBy).toEqual("Title");
        searchPage.instance().onSearchParameterClick({
            target: {
                dataset: {
                    parameter: "Genre"
                }
            }
        });
        expect(searchPage.props().searchBy).toEqual("Genre");
        searchPage.instance().onSearchModeClick();
        expect(searchPage.props().chosenFilm).toEqual(null);
        expect(searchPage.props().sortBy).toEqual("rating");
        searchPage.instance().chooseFilm({
            target: {
                dataset: {
                    filmId: ""
                }
            }
        });
        expect(searchPage.props().chosenFilm).toEqual(null);
        searchPage.instance().chooseFilm({
            target: {
                dataset: {
                    filmId: 2222
                }
            }
        });
        expect(searchPage.props().chosenFilm).toEqual({
            id: 2222,
            release_date: "2020-05-01",
            title: "Guardians of the Galaxy Vol. 3",
            vote_average: 8,
            runtime: 100,
            tagline: "Tag line",
            overview: "The third film based on Marvel's Guardians of the Galaxy.",
            poster_path: "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
            genres: ['Drama', 'Adventure']
        });
        searchPage.instance().chooseFilm({
            target: {
                dataset: {
                    filmId: "????"
                }
            }
        });
        expect(searchPage.props().chosenFilm).toEqual(null);
    });
})
