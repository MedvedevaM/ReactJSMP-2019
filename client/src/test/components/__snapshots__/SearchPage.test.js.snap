// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search page rendering and functionality check radio buttons rendering 1`] = `
<Fragment>
  <FilmSearch
    chosenFilm={null}
    fetchFilms={
      [MockFunction] {
        "calls": Array [
          Array [
            "http://reactjs-cdp.herokuapp.com/movies",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    films={
      Array [
        Object {
          "genres": Array [
            "Drama",
            "Adventure",
          ],
          "id": 1111,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "runtime": 100,
          "tagline": "Tag line",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 8,
        },
        Object {
          "genres": Array [
            "Drama",
            "Adventure",
          ],
          "id": 2222,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "runtime": 100,
          "tagline": "Tag line",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 8,
        },
      ]
    }
    getSortedFilms={[MockFunction]}
    onSearchModeClick={[Function]}
    onSearchParameterClick={[Function]}
    quantityOfFilms={2}
    searchBy="Title"
    searchFilms={[Function]}
    searchParameter={[MockFunction]}
    setChosenFilm={[MockFunction]}
    setSearchParameter={[MockFunction]}
    setSortParameter={[MockFunction]}
    sortBy="release date"
  />
  <Connect(FilmsContainer)
    chooseFilm={[Function]}
    chosenFilm={null}
    fetchFilms={
      [MockFunction] {
        "calls": Array [
          Array [
            "http://reactjs-cdp.herokuapp.com/movies",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    films={
      Array [
        Object {
          "genres": Array [
            "Drama",
            "Adventure",
          ],
          "id": 1111,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "runtime": 100,
          "tagline": "Tag line",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 8,
        },
        Object {
          "genres": Array [
            "Drama",
            "Adventure",
          ],
          "id": 2222,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "runtime": 100,
          "tagline": "Tag line",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 8,
        },
      ]
    }
    getSortedFilms={[MockFunction]}
    onSortParameterClick={[Function]}
    quantityOfFilms={2}
    searchBy="Title"
    searchParameter={[MockFunction]}
    setChosenFilm={[MockFunction]}
    setSearchParameter={[MockFunction]}
    setSortParameter={[MockFunction]}
    sortBy="release date"
  />
  <footer>
    <div>
      <p
        className="logo common-color-1"
      >
        netflixroulette
      </p>
    </div>
  </footer>
</Fragment>
`;
